diff --git a/src/hunspell/hunspell.cxx b/src/hunspell/hunspell.cxx
index a9b261a..93fe796 100644
--- a/src/hunspell/hunspell.cxx
+++ b/src/hunspell/hunspell.cxx
@@ -1670,6 +1670,13 @@ int Hunspell::get_langnum() const
    return langnum;
 }
 
+int Hunspell::input_conv(const char * word, char * dest)
+{
+  RepList * rl = (pAMgr) ? pAMgr->get_iconvtable() : NULL;
+  return (rl && rl->conv(word, dest));
+}
+
+
 // return the beginning of the element (attr == NULL) or the attribute
 const char * Hunspell::get_xml_pos(const char * s, const char * attr)
 {
diff --git a/src/hunspell/hunspell.hxx b/src/hunspell/hunspell.hxx
index 9b6c388..4d6984b 100644
--- a/src/hunspell/hunspell.hxx
+++ b/src/hunspell/hunspell.hxx
@@ -131,6 +131,9 @@ public:
   const char * get_version();
 
   int get_langnum() const;
+
+  /* need for putdic */
+  int input_conv(const char * word, char * dest);
   
   /* experimental and deprecated functions */
 
diff --git a/src/tools/hunspell.cxx b/src/tools/hunspell.cxx
index 1a65d0d..ec31bfe 100644
--- a/src/tools/hunspell.cxx
+++ b/src/tools/hunspell.cxx
@@ -402,12 +402,17 @@ void log(char * message)
 int putdic(char * word, Hunspell * pMS)
 {
     char * w;
-
+    char buf[MAXLNLEN];
+    
     word = chenc(word, ui_enc, dic_enc[0]);
+
+    if(pMS->input_conv(word, buf)) word = buf;
+    
+    int ret;
     
     if (((w = strstr(word + 1, "/")) == NULL)) {
-        if (*word == '*') return pMS->remove(word + 1);
-	else return pMS->add(word);
+        if (*word == '*') ret =  pMS->remove(word + 1);
+	else ret = pMS->add(word);
     } else {
 	char c;
 	int ret;
@@ -419,8 +424,8 @@ int putdic(char * word, Hunspell * pMS)
 	    ret = pMS->add_with_affix(word, w + 1); // word/pattern
 	}
 	*w = c;
-	return ret;
     }
+    return ret;
 }
 
 void load_privdic(char * filename, Hunspell * pMS) 
